### PR TITLE
QC-1181: Multi select power and styles

### PR DESCRIPTION
**Pull request by sirJiggles**
### What

For our story we need to have the ability to have multi select so on looking at the plugin we have already "react-select" I noticed we basically already have it built in. So I just enabled the prop to be passed, handled the on change / value and styled her up as per the design.

For ref this is the story:

https://qualifyze.atlassian.net/browse/QC-1121

and this is the design:

https://www.figma.com/file/iMYDLEBL492Dj8nj3bZMuO/Admin?node-id=2816%3A7361

Also note there are a few new types of select components now, select, multi select and the creatable for both. The creatable is a variant of each that will allow the user to create new entries in the select. 

### How she looks:

Screenshot 2021-08-02 at 16 18 06
Screenshot 2021-08-02 at 16 18 17

---
This change is [Reviewable](https://reviewable.io/reviews/qualifyze/design-system/102)

### CODE REVIEW

**Review by vubogovich — COMMENTED** [review]
Good start! I hope it supports the options provided dynamically but how do we display the "loading" state?
Is the user able to type something new as an option? two new options?
It would be also nice to be able to display secondary names as discussed with Boris.

> *[Reviewable](https://reviewable.io/reviews/qualifyze/design-system/102#-:-Mfs7IN9CW4rSz6jmi_t:bibd426)* status: 0 of 3 files reviewed, all discussions resolved

**Review by gregoralbrecht — CHANGES_REQUESTED** [review]
Always happy to see our DS grow :heart_eyes: 

We're trying to keep API surfaces somehow small, so I believe we should make this a separate `MultiSelectField` component instead of extending the `SelectField` API. 

> Reviewed 1 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/qualifyze/design-system/102#-:-Mfw7Ka80WpXR0rmUr1_:b3uen8f)* status: 1 of 3 files reviewed, 4 unresolved discussions (waiting on @sirJiggles)

*[a discussion](https://reviewable.io/reviews/qualifyze/design-system/102#-MfwERExAv8q3adpeBOR:-MfwERExAv8q3adpeBOS:bi1m09t) (no related file):*
I noticed that the selected options are reordered when you select them, which is a bit confusing. In the Storybook this means:
1. Select `Ascorbic` first -> It shows `[Ascorbic]` :heavy_check_mark: 
2. Select `Doxylamine` second -> It shows `[Doxylamine, Ascorbic]`

So if we think about a list of selected options, all will be moved when adding another option if the user doesn't select them in order. I can't tell why it happens but we should avoid it to not confuse anyone.

---
*[a discussion](https://reviewable.io/reviews/qualifyze/design-system/102#-MfwF3hf7GNlf42tLV9v:-MfwF3hf7GNlf42tLV9w:bh69bi7) (no related file):*
A strong quality of `react-select` is that we can customize the components it uses. In this case, it's pretty obvious that the clear indicator is the default one because a) it doesn't have our usual hover indicators like `cursor: pointer` and a changing color, and b) it doesn't use our icons. 

Not fixing b) is totally fine because we didn't do this for the rest of the `SelectField` component either. So it's somehow consistent.

I think we can easily fix a), though. And there's another benefit to that: We can ensure the click target is at least 44px by 44px in size, which is nice for mobile.

`react-select` allows you to replace components with the `components` prop like this: https://codesandbox.io/s/muddy-thunder-heyr0?file=/src/App.js

You can take a look at the icon buttons here and pretty much copy&paste most of it: https://github.com/Qualifyze/design-system/blob/main/src/components/FileBox/PrimaryAction.jsx

tl;dr Let's...
1. ...use the cross icon from `react-select`
2. ...add `cursor: pointer` to the `MultiValueRemove`
3. ...make sure the click area of `MultiValueRemove` is at least 44x44px

Let me know if any of this is confusing or cryptic :)

---
___
*[src/components/SelectField/baseTheme.js, line 27 at r1](https://reviewable.io/reviews/qualifyze/design-system/102#-Mfw7uTD77I2ZECGt6yQ:-Mfw7uTD77I2ZECGt6yR:b-2vhos5) ([raw file](https://github.com/qualifyze/design-system/blob/18e54b87165ddefc7123eb7b07c70457c5f29159/src/components/SelectField/baseTheme.js#L27)):*
> ```JavaScript
>       backgroundColor: theme.colors.grey[300],
>       color: theme.colors.grey[800],
>       borderRadius: theme.radii[3],
> ```

IMHO this looks a bit odd because it doesn't match with the border-radius of the select input. Let's go with `2` for now, we can easily change it later if we want to
___
*[src/components/SelectField/index.jsx, line 69 at r1](https://reviewable.io/reviews/qualifyze/design-system/102#-Mfw8bINC2MwRtzcF3yz:-Mfw8bINC2MwRtzcF3z-:b-wl6h7m) ([raw file](https://github.com/qualifyze/design-system/blob/18e54b87165ddefc7123eb7b07c70457c5f29159/src/components/SelectField/index.jsx#L69)):*
> ```JavaScript
>   size,
>   menuPlacement,
>   ...props
> ```

Spreading all props is usually not the best idea in a component like this because it makes the API unnecessarily large. It also makes it difficult to ensure everything works correctly when refactoring. For instance, a consumer might pass some test id or something else and then the underlying markup changes. So I believe it's best to avoid it and pass props explicitly.

**Review by gregoralbrecht — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/qualifyze/design-system/102)* status: 1 of 3 files reviewed, 4 unresolved discussions (waiting on @sirJiggles)

*[a discussion](https://reviewable.io/reviews/qualifyze/design-system/102#-MfwF3hf7GNlf42tLV9v:-MfwQu_0BqCD_isC2yu5:bszrbxv) (no related file):*
<details><summary><i>Previously, gregoralbrecht (Gregor Albrecht) wrote…</i></summary><blockquote>

A strong quality of `react-select` is that we can customize the components it uses. In this case, it's pretty obvious that the clear indicator is the default one because a) it doesn't have our usual hover indicators like `cursor: pointer` and a changing color, and b) it doesn't use our icons. 

Not fixing b) is totally fine because we didn't do this for the rest of the `SelectField` component either. So it's somehow consistent.

I think we can easily fix a), though. And there's another benefit to that: We can ensure the click target is at least 44px by 44px in size, which is nice for mobile.

`react-select` allows you to replace components with the `components` prop like this: https://codesandbox.io/s/muddy-thunder-heyr0?file=/src/App.js

You can take a look at the icon buttons here and pretty much copy&paste most of it: https://github.com/Qualifyze/design-system/blob/main/src/components/FileBox/PrimaryAction.jsx

tl;dr Let's...
1. ...use the cross icon from `react-select`
2. ...add `cursor: pointer` to the `MultiValueRemove`
3. ...make sure the click area of `MultiValueRemove` is at least 44x44px

Let me know if any of this is confusing or cryptic :)
</blockquote></details>

FYI I opened a PR here to use a custom icon here: https://github.com/Qualifyze/design-system/pull/103

---

**Review by sirJiggles — COMMENTED** [review]
Sure thing, I kept them in the same file though as there is like 300 lines of stuff that would need to be duplicated and I think as they are kind of the same component but one with another prop it is ok to assume *for now at least. That it would be worse to have them deviate on these styles etc than have shared code. Let me know what you think

> *[Reviewable](https://reviewable.io/reviews/qualifyze/design-system/102#-:-Mg4mwq6B2NT1236SpXT:b-u3wpt3)* status: 0 of 4 files reviewed, 4 unresolved discussions (waiting on @gregoralbrecht)

*[a discussion](https://reviewable.io/reviews/qualifyze/design-system/102#-MfwERExAv8q3adpeBOR:-Mg4mt8FCDl1PSaqBXzZ:b-rx548k) (no related file):*
<details><summary><i>Previously, gregoralbrecht (Gregor Albrecht) wrote…</i></summary><blockquote>

I noticed that the selected options are reordered when you select them, which is a bit confusing. In the Storybook this means:
1. Select `Ascorbic` first -> It shows `[Ascorbic]` :heavy_check_mark: 
2. Select `Doxylamine` second -> It shows `[Doxylamine, Ascorbic]`

So if we think about a list of selected options, all will be moved when adding another option if the user doesn't select them in order. I can't tell why it happens but we should avoid it to not confuse anyone.
</blockquote></details>

Nice spot, this is done :thumbsup: The issue is to do with the way we set the "value" field on the select, which is needed for things like the reset button. I have resolved it now, the order should stay, there is a also a comment for the next traveler

---
*[a discussion](https://reviewable.io/reviews/qualifyze/design-system/102#-MfwF3hf7GNlf42tLV9v:-Mg54gw9A6Lng8Inlnj0:b-ehnwvb) (no related file):*
<details><summary><i>Previously, gregoralbrecht (Gregor Albrecht) wrote…</i></summary><blockquote>

FYI I opened a PR here to use a custom icon here: https://github.com/Qualifyze/design-system/pull/103
</blockquote></details>

All done, and really nice that we can do this. thanks for taking the time to show me :thumbsup: 

---
___
*[src/components/SelectField/baseTheme.js, line 27 at r1](https://reviewable.io/reviews/qualifyze/design-system/102#-Mfw7uTD77I2ZECGt6yQ:-Mg4mRx19-sJcucXeoa2:bwyuh9j) ([raw file](https://github.com/qualifyze/design-system/blob/18e54b87165ddefc7123eb7b07c70457c5f29159/src/components/SelectField/baseTheme.js#L27)):*
<details><summary><i>Previously, gregoralbrecht (Gregor Albrecht) wrote…</i></summary><blockquote>

IMHO this looks a bit odd because it doesn't match with the border-radius of the select input. Let's go with `2` for now, we can easily change it later if we want to
</blockquote></details>

Done. was going form figma. But agree lets keep her consistent
___
*[src/components/SelectField/index.jsx, line 69 at r1](https://reviewable.io/reviews/qualifyze/design-system/102#-Mfw8bINC2MwRtzcF3yz:-Mg4mR6c4UKtJpaKacjo:b88ecsv) ([raw file](https://github.com/qualifyze/design-system/blob/18e54b87165ddefc7123eb7b07c70457c5f29159/src/components/SelectField/index.jsx#L69)):*
<details><summary><i>Previously, gregoralbrecht (Gregor Albrecht) wrote…</i></summary><blockquote>

Spreading all props is usually not the best idea in a component like this because it makes the API unnecessarily large. It also makes it difficult to ensure everything works correctly when refactoring. For instance, a consumer might pass some test id or something else and then the underlying markup changes. So I believe it's best to avoid it and pass props explicitly.
</blockquote></details>

We have removed this now, and thanks for the reminder about the dangers of this spreading :thumbsup:

**Review by sirJiggles — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/qualifyze/design-system/102)* status: 0 of 4 files reviewed, 4 unresolved discussions (waiting on @gregoralbrecht)

*[a discussion](https://reviewable.io/reviews/qualifyze/design-system/102#-MfwF3hf7GNlf42tLV9v:-Mg55HxBA2ioDBKSlHUV:b-896fix) (no related file):*
<details><summary><i>Previously, sirJiggles (Gareth Fuller) wrote…</i></summary><blockquote>

All done, and really nice that we can do this. thanks for taking the time to show me :thumbsup: 
</blockquote></details>

Done.

---

**Review by sirJiggles — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/qualifyze/design-system/102)* status: 0 of 4 files reviewed, 4 unresolved discussions (waiting on @gregoralbrecht)

___
*[src/components/SelectField/baseTheme.js, line 27 at r1](https://reviewable.io/reviews/qualifyze/design-system/102#-Mfw7uTD77I2ZECGt6yQ:-Mg55JOW5Sinp5Dw5m6Z:b-896fix) ([raw file](https://github.com/qualifyze/design-system/blob/18e54b87165ddefc7123eb7b07c70457c5f29159/src/components/SelectField/baseTheme.js#L27)):*
<details><summary><i>Previously, sirJiggles (Gareth Fuller) wrote…</i></summary><blockquote>

Done. was going form figma. But agree lets keep her consistent
</blockquote></details>

Done.
___
*[src/components/SelectField/index.jsx, line 69 at r1](https://reviewable.io/reviews/qualifyze/design-system/102#-Mfw8bINC2MwRtzcF3yz:-Mg55J7W3F5cbPq-cppf:b-896fix) ([raw file](https://github.com/qualifyze/design-system/blob/18e54b87165ddefc7123eb7b07c70457c5f29159/src/components/SelectField/index.jsx#L69)):*
<details><summary><i>Previously, sirJiggles (Gareth Fuller) wrote…</i></summary><blockquote>

We have removed this now, and thanks for the reminder about the dangers of this spreading :thumbsup: 
</blockquote></details>

Done.

**Review by sirJiggles — COMMENTED** [review]
@vubogovich, sorry I could not reply direct to your comment but I will now add this: https://react-select.com/creatable thanks for the reminder I will make another story where users can see how to make it also "creatable" to cover this use case :thumbsup: 

> *[Reviewable](https://reviewable.io/reviews/qualifyze/design-system/102#-:-Mg56-2MCQzMkHT6uFQU:b-jwmao5)* status: 0 of 4 files reviewed, 4 unresolved discussions (waiting on @gregoralbrecht)

**Review by gregoralbrecht — CHANGES_REQUESTED** [review]
I still think we can have them be separate components, the duplication doesn't hurt.

> Reviewed 2 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/qualifyze/design-system/102#-:-Mg564KrAlfri_8SK6-P:bsqr9gq)* status: 2 of 4 files reviewed, 1 unresolved discussion (waiting on @sirJiggles)

*[a discussion](https://reviewable.io/reviews/qualifyze/design-system/102#-MfwERExAv8q3adpeBOR:-Mg57GzI8vbRSC7u6ter:b3wib39) (no related file):*
<details><summary><i>Previously, sirJiggles (Gareth Fuller) wrote…</i></summary><blockquote>

Nice spot, this is done :thumbsup: The issue is to do with the way we set the "value" field on the select, which is needed for things like the reset button. I have resolved it now, the order should stay, there is a also a comment for the next traveler
</blockquote></details>

Nice thanks! Works perfectly well :100: 

---
*[a discussion](https://reviewable.io/reviews/qualifyze/design-system/102#-MfwF3hf7GNlf42tLV9v:-Mg57lSf3wPtFhfFv0W9:bxnykrg) (no related file):*
<details><summary><i>Previously, sirJiggles (Gareth Fuller) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Thanks for adding a custom component! There are some styling issues (input gets bigger when something is selected and the multi value label has a lot of padding, etc.) I will look into it and open a PR to fix this

---

**Review by gregoralbrecht — CHANGES_REQUESTED** [review]
As discussed in Slack, let's have separate PRs for the changes. Also, I'm still convinced it makes sense to have everything be separate folders. If we want to avoid duplicating things like the base theme 4 times we can put it into `SelectField` and import it from there. These components are obvious related so a little interdependence should be fine

> *[Reviewable](https://reviewable.io/reviews/qualifyze/design-system/102#-:-Mg7Eud3DDmUQlIJ6IYz:b-a3yvzb)* status: 1 of 9 files reviewed, 1 unresolved discussion (waiting on @gregoralbrecht, @pablosci, @sirJiggles, and @vubogovich)

**Review by vubogovich — COMMENTED** [review]
> Reviewed 1 of 4 files at r2, 8 of 8 files at r3.
> *[Reviewable](https://reviewable.io/reviews/qualifyze/design-system/102)* status: all files reviewed, 2 unresolved discussions (waiting on @pablosci and @sirJiggles)

*[a discussion](https://reviewable.io/reviews/qualifyze/design-system/102#-Mg7iZsxCkDZDjVG0lhA:-Mg7iZsxCkDZDjVG0lhB:bf55nvg) (no related file):*
Started a new discussion so it's easier to reply.
Multiselect creatable works very nice! The open questions are:
- does it support loading the options dynamically? how do we display the "loading" state?
- can we display secondary product names in the options like "Object Tags" on https://angular-ui.github.io/ui-select/demo-tagging.html ?

---

**Review by sirJiggles — COMMENTED** [review]
Sure thing I will make another PR for each component and break them down into each folder, it also solves a rollup issue I have with default exports and tree shaking :thumbsup:  I will close this PR in a sec

> *[Reviewable](https://reviewable.io/reviews/qualifyze/design-system/102#-:-Mg9qTUY0WCwKLYAbVSW:bbd1r13)* status: all files reviewed, 2 unresolved discussions (waiting on @gregoralbrecht, @pablosci, and @vubogovich)

*[a discussion](https://reviewable.io/reviews/qualifyze/design-system/102#-MfwF3hf7GNlf42tLV9v:-Mg9pfLw9eUSF3-CeSWh:b-896fix) (no related file):*
<details><summary><i>Previously, gregoralbrecht (Gregor Albrecht) wrote…</i></summary><blockquote>

Thanks for adding a custom component! There are some styling issues (input gets bigger when something is selected and the multi value label has a lot of padding, etc.) I will look into it and open a PR to fix this
</blockquote></details>

Done.

---
*[a discussion](https://reviewable.io/reviews/qualifyze/design-system/102#-Mg7iZsxCkDZDjVG0lhA:-Mg9pgmm9K-0rJOpeYC1:b-ad2jvd) (no related file):*
<details><summary><i>Previously, vubogovich (Vladimir Ubogovich) wrote…</i></summary><blockquote>

Started a new discussion so it's easier to reply.
Multiselect creatable works very nice! The open questions are:
- does it support loading the options dynamically? how do we display the "loading" state?
- can we display secondary product names in the options like "Object Tags" on https://angular-ui.github.io/ui-select/demo-tagging.html ?
</blockquote></details>

I will look at the loading state of dynamic inputs in another PR :thumbsup: 

with regards to secondary options, you could probably rig it up to show something like that, but to be honest I would not. We have the international names of the products we can always show and the searching of each of the other props should be done in algolia. This component should just be about searching one dimension in my opinion.

it makes it easier to understand for the user and easier for us to make something that can be used in other contexts.

If we do however decide we want to deviate that much from react select then I would say we defo need another PR for this where we will apply these sort of customisations, not sure if we would want something like that in the ds.

Anyhow I am going to break all these things up this morning into other PR's and close this one

---
